ocaml-lsp-server/src/ocaml_lsp_server.ml

```diff
@@ -511,6 +516,46 @@ let definition_query (state : State.t) uri position merlin_request =
   let result = location_of_merlin_loc uri result in
   Ok (result, state)
 
+(** [switch_impl_intf path] finds a filepath for a file to switch to based on
+    [path]'s extension or a list of possible filenames to create
+
+    @return a list of single filepath to switch to if it already exists or a
+    list of filepaths to create a file *)
+let switch_impl_intf (state : State.t) (filepath : string) :
+    (Json.t, Jsonrpc.Response.Error.t) result =
+  let uri = Uri.of_path filepath in
+  let open Result.O in
+  let+ doc = Document_store.get state.store uri in
+  let ml, mli, re, rei, mll, mly = ("ml", "mli", "re", "rei", "mll", "mly") in
+  let path_without_extension = Filename.remove_extension filepath ^ "." in
+  let find_switch (exts : string list) =
```

Comment:
This logic is slightly backwards. It should be something like this:

1) Get all the candidates
2) If there's only a single candidate, choose it
3) If there's more than one candidate, pick the one that already exists.

Reply:
Let's have the following logic: 

```
if file is <extension> file then
  (case 1) if only one candidate exists then open that file else 
  (case 2) if several candidates exist then open a dialog box for user to pick which to open (or open all? - useful) else
  (case 3) if no candidate exists then <discussed below>
```

For cases 1 and 2, the behavior for files is clear, let's discuss case 3 - no candidate exists.

ml: create mli file directly
mli: create ml file directly
re: create rei file directly
rei: create re file directly
mll: offer to run `ocamllex <given filepath>` or do nothing (?)
mly: offer to choose from menhir and ocamlyacc and run or do nothing (?)

Wdyt?

Reply:
How about this:

If no candidates exist, then we present all possibilities to the user. The user can then pick and save this file. Consequently, the next invocations will no longer need to prompt the user. That seems simple and also consistent for all file types.